x/fee_grant/internal/types/periodic_fee.go

```diff
@@ -0,0 +1,113 @@
+package types
+
+import (
+	"time"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/x/fee_grant/exported"
+)
+
+// PeriodicFeeAllowance extends FeeAllowance to allow for both a maximum cap,
+// as well as a limit per time period.
+type PeriodicFeeAllowance struct {
+	Basic BasicFeeAllowance
+
+	// Period is the duration of one period
+	Period Duration
+	// PeriodSpendLimit is the maximum amount of tokens to be spent in this period
+	PeriodSpendLimit sdk.Coins
+
+	// PeriodCanSpend is how much is available until PeriodReset
+	PeriodCanSpend sdk.Coins
+
+	// PeriodRest is when the PeriodCanSpend is updated
+	PeriodReset ExpiresAt
+}
+
+var _ exported.FeeAllowance = (*PeriodicFeeAllowance)(nil)
+
+// Accept implements FeeAllowance and deducts the fees from the SpendLimit if possible
+func (a *PeriodicFeeAllowance) Accept(fee sdk.Coins, blockTime time.Time, blockHeight int64) (remove bool, err error) {
+	if a.Basic.Expiration.IsExpired(blockTime, blockHeight) {
+		return true, ErrFeeLimitExpired()
+	}
+
+	a.TryResetPeriod(blockTime, blockHeight)
+
+	// deduct from both the current period and the max amount
+	var isNeg bool
+	a.PeriodCanSpend, isNeg = a.PeriodCanSpend.SafeSub(fee)
+	if isNeg {
+		return false, ErrFeeLimitExceeded()
+	}
+	a.Basic.SpendLimit, isNeg = a.Basic.SpendLimit.SafeSub(fee)
+	if isNeg {
+		return false, ErrFeeLimitExceeded()
+	}
+
+	return a.Basic.SpendLimit.IsZero(), nil
+}
+
+// TryResetPeriod will reset the period if we hit the conditions
+func (a *PeriodicFeeAllowance) TryResetPeriod(blockTime time.Time, blockHeight int64) {
```

Comment:
This needs a much more elaborate explanation of the business logic and all the steps in the godoc IMHO.